File: src/main/java/main/java/SearchProcess.java
package main.java;

import java.util.ArrayList;
import java.util.HashMap;

public class SearchProcess {

    public ArrayList<String> middleProcess(ArrayList<String> orQueries, ArrayList<String> andQueries, ArrayList<String> excludeQueries, HashMap<String, ArrayList<String>> dictionary){
        ArrayList <String> result = new ArrayList<>();
        for(String word : orQueries){
            if(!dictionary.containsKey(word.substring(1)))
                continue;
            //result.addAll(dictionary.get(word.substring(1)));
            for(String doc : dictionary.get(word.substring(1))){
                if(!result.contains(doc)){
                    result.add(doc);
                }
            }
        }

        if(result.isEmpty()){
            for(String word : andQueries){
                if(dictionary.containsKey(word)){
                    for(String doc : dictionary.get(word.substring(1))){
                        if(!result.contains(doc)){
                            result.add(doc);
                        }
                    }
                    break;
                }
            }
        }

        for(String word : andQueries){
            if(!dictionary.containsKey(word)) {
                result.clear();
                break;
            }
            result.removeIf(document -> !dictionary.get(word).contains(document));
        }

        for(String word : excludeQueries){
            if(!dictionary.containsKey(word.substring(1)))
                continue;
            for(String document : dictionary.get(word.substring(1))){
                result.remove(document);
            }
        }
        return result;
    }
}

File: phase5/SampleLibrary/SampleLibrary/SearchProcess.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace SampleLibrary
{
    public class SearchProcess
    {
        public ArrayList MiddleProcess(ArrayList orQueries, ArrayList andQueries, ArrayList exQueries,
            Dictionary<string, ArrayList> dictionary)
        {
            ArrayList result = new ArrayList();
            foreach (string word in orQueries)
            {
                if (!dictionary.ContainsKey(word.Substring(1)))
                {
                    continue;
                }

                foreach (string doc in dictionary[word.Substring(1)])
                {
                    if (!result.Contains(doc))
                        result.Add(doc);
                }
            }

            if (result.Count == 0)
            {
                foreach (string word in andQueries)
                {
                    if (dictionary.ContainsKey(word))
                    {
                        foreach (string doc in dictionary[word.Substring(1)])
                        {
                            if (!result.Contains(doc))
                            {
                                result.Add(doc);
                            }
                        }
                        break;
                    }
                }
            }

            foreach (string word in andQueries)
            {
                if (!dictionary.ContainsKey(word))
                {
                    result.Clear();
                    break;
                }

                foreach (string doc in result)
                {
                    if (!dictionary[word.Substring(1)].Contains(doc))
                    {
                        result.Remove(doc);
                    }
                }
            }
            foreach(string word in exQueries){
                if(!dictionary.ContainsKey(word.Substring(1)))
                    continue;
                foreach(string document in dictionary[word.Substring(1)]){
                    result.Remove(document);
                }
            }
            return result;
        }
    }
}